### PR TITLE
Add lists to vortex

### PR DESCRIPTION
**Pull request by joseph-isaacs**
I add a simple implementation of lists based off varbin. I hope to unify parts of both of these.

### CODE REVIEW

File: vortex-array/src/array/list/mod.rs [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+    offset_dtype: DType,
+}
+
+impl Display for ListMetadata {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "ListMetadata")
+    }
+}
+
+#[allow(dead_code)]
```
Comment (a10y) [inline vortex-array/src/array/list/mod.rs]: do we need this

File: vortex-array/src/arrow/array.rs [hunk 2]
```diff
@@ -177,6 +178,20 @@ impl FromArrowArray<&ArrowStructArray> for ArrayData {
     }
 }
 
+impl<O: OffsetSizeTrait + NativePType> FromArrowArray<&GenericListArray<O>> for ArrayData {
+    fn from_arrow(value: &GenericListArray<O>, nullable: bool) -> Self {
+        // let elements = Self::from_arrow(value.values(), value.values().is_nullable());
```
Comment (a10y) [inline vortex-array/src/arrow/array.rs]: get rid?

File: vortex-array/src/canonical.rs [hunk 3]
```diff
@@ -220,6 +231,18 @@ fn struct_to_arrow(struct_array: StructArray) -> VortexResult<ArrayRef> {
...
+fn list_to_arrow(list: ListArray) -> VortexResult<ArrayRef> {
+    Ok(Arc::new(arrow_array::ListArray::try_new(
+        FieldRef::new(Field::new_list_field(
+            infer_data_type(list.elements().dtype())?,
+            list.validity().nullability().into(),
+        )),
+        as_offset_buffer::<i32>(list.offsets().into_primitive()?),
+        list.elements().into_canonical()?.into_arrow()?,
+        list.logical_validity().to_null_buffer()?,
+    )?) as ArrayRef)
```
Comment (a10y) [inline vortex-array/src/canonical.rs]: we should return a `LargeListArray` to arrow if the indices have u64 PType

File: vortex-dtype/src/dtype.rs [hunk 4]
```diff
@@ -138,6 +138,14 @@ impl DType {
             _ => None,
         }
     }
+
+    /// Get the inner dtype if `self` is a `ListDType`, otherwise `None`
+    pub fn as_list(&self) -> Option<(&DType, &Nullability)> {
```
Comment (a10y) [inline vortex-dtype/src/dtype.rs]: do we need to return outer nullability here? e.g. as_struct doesn't do this

Comment (a10y) [reply vortex-dtype/src/dtype.rs]: it looks like everywhere this function is called it's `let (dtype, _) = ` so perhaps we can remove it

Comment (joseph-isaacs) [reply vortex-dtype/src/dtype.rs]: I have removed it, but it does seem strange that as_struct, doesn't I think it is that vortex top level compute functions apply nullability.

Comment (joseph-isaacs) [reply vortex-dtype/src/dtype.rs]: However it might be nice as a reminder that it exists.

File: vortex-array/src/array/chunked/canonical.rs [hunk 5]
```diff
@@ -88,9 +89,11 @@ pub(crate) fn try_canonicalize_chunks(
             )))
         }
 
-        // TODO(aduffy): better list support
         DType::List(..) => {
-            todo!()
+            // TODO: improve performance
```
Comment (gatesn) [inline vortex-array/src/array/chunked/canonical.rs]: Make sure you use the `TODO(username):` pattern so we can ask the author even modulo new git commits

Comment (gatesn) [reply vortex-array/src/array/chunked/canonical.rs]: But also why is this comment here? This function is fast! It only calls another function

File: vortex-array/src/array/chunked/canonical.rs [hunk 6]
```diff
@@ -117,6 +120,41 @@ pub(crate) fn try_canonicalize_chunks(
...
+    let len: usize = chunks.iter().map(|c| c.len()).sum();
+    let mut offsets = Vec::with_capacity(len + 1);
+    offsets.push(0);
+    let mut elements = Vec::new();
+
+    for chunk in chunks {
+        let chunk = chunk.clone().into_list()?;
+        // TODO: handle i32 offsets if they fit.
+        let offsets_arr = try_cast(
+            chunk.offsets().into_primitive()?.as_ref(),
```
Comment (gatesn) [inline vortex-array/src/array/chunked/canonical.rs]: Do you need to into_primitive before the cast?

File: vortex-array/src/array/chunked/canonical.rs [hunk 7]
```diff
@@ -117,6 +120,41 @@ pub(crate) fn try_canonicalize_chunks(
...
+            &DType::Primitive(PType::I64, Nullability::NonNullable),
+        )?
+        .into_primitive()?;
+
+        let first_offset_value: usize = usize::try_from(&scalar_at(offsets_arr.as_ref(), 0)?)?;
+        elements.push(chunk.elements());
+
+        let adjustment_from_previous = *offsets
+            .last()
+            .ok_or_else(|| vortex_err!("VarBinArray offsets must have at least one element"))?;
```
Comment (gatesn) [inline vortex-array/src/array/chunked/canonical.rs]: This isn't VarBinArray

File: vortex-array/src/array/list/compute/mod.rs [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+        Ok(ListArray::try_new(
+            array.elements(),
+            slice(array.offsets(), start, stop + 1)?,
+            array.validity().slice(start, stop)?,
+        )?
+        .into_array())
+    }
+}
+
+fn list_scalar_of(array: ArrayData) -> VortexResult<Scalar> {
```
Comment (gatesn) [inline vortex-array/src/array/list/compute/mod.rs]: Don't pull out functions that are only called once.

File: vortex-array/src/array/list/mod.rs [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+    IntoArrayVariant, IntoCanonical,
+};
+
+impl_encoding!("vortex.list", ids::LIST, List);
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct ListMetadata {
+    validity: ValidityMetadata,
+    element_len: usize,
+    offset_dtype: DType,
```
Comment (gatesn) [inline vortex-array/src/array/list/mod.rs]: Can this be PType?

File: vortex-array/src/array/list/mod.rs [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+    impl_encoding, ArrayDType, ArrayData, ArrayLen, ArrayTrait, Canonical, IntoArrayData,
+    IntoArrayVariant, IntoCanonical,
+};
+
+impl_encoding!("vortex.list", ids::LIST, List);
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct ListMetadata {
+    validity: ValidityMetadata,
+    element_len: usize,
```
Comment (gatesn) [inline vortex-array/src/array/list/mod.rs]:
```suggestion
    elements_len: usize,
```

File: vortex-array/src/array/list/mod.rs [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+                "Expected offsets to be an non-nullable integer type, got {:?}",
+                offsets.dtype()
+            );
+        }
+
+        if offsets.is_empty() {
+            vortex_bail!("Offsets must have at least one element, [0] for an empty list");
+        }
+
+        let Some(min_value) = offsets.statistics().compute_as_cast::<i64>(Stat::Min) else {
```
Comment (gatesn) [inline vortex-array/src/array/list/mod.rs]: You shouldn't compute a statistic here, constructing an array should be relatively cheap. You can force a signed integer offsets dtype if you like

File: vortex-array/src/array/list/mod.rs [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+            unreachable!("Array must have min value");
+        };
+        if min_value < 0 {
+            vortex_bail!(
+                "Expected smallest value in offsets array must be, non-negative {}",
+                min_value
+            );
+        }
+
+        if offsets.dtype().is_signed_int() {
```
Comment (gatesn) [inline vortex-array/src/array/list/mod.rs]: These two code paths are too duplicative

File: vortex-array/src/array/list/mod.rs [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+                vortex_bail!(
+                    "Expected smallest value in offsets array must be, less than final idx {}",
+                    final_idx
+                );
+            }
+        } else {
+            unreachable!("Offsets are integers");
+        }
+
+        let is_sorted = offsets.statistics().compute_is_sorted();
```
Comment (gatesn) [inline vortex-array/src/array/list/mod.rs]: Again, don't compute the statistic.

File: vortex-array/src/arrow/array.rs [hunk 14]
```diff
@@ -177,6 +178,19 @@ impl FromArrowArray<&ArrowStructArray> for ArrayData {
     }
 }
 
+impl<O: OffsetSizeTrait + NativePType> FromArrowArray<&GenericListArray<O>> for ArrayData {
+    fn from_arrow(value: &GenericListArray<O>, nullable: bool) -> Self {
+        ListArray::try_new(
+            Self::from_arrow(value.values().clone(), value.values().is_nullable()),
+            // offsets are always non-nullable
```
Comment (gatesn) [inline vortex-array/src/arrow/array.rs:185]: Still use the `from_arrow` constructor though. We should remove the other one tbh

File: vortex-array/src/encoding/mod.rs [hunk 15]
```diff
@@ -122,10 +122,10 @@ pub mod ids {
     pub const SPARSE: u16 = 8;
     pub const CONSTANT: u16 = 9;
     pub const CHUNKED: u16 = 10;
+    pub(crate) const LIST: u16 = 11;
```
Comment (gatesn) [inline vortex-array/src/encoding/mod.rs]: Why is this pub(crate) and the others aren't?

File: vortex-dtype/src/dtype.rs [hunk 16]
```diff
@@ -138,6 +138,14 @@ impl DType {
             _ => None,
         }
     }
+
+    /// Get the inner dtype if `self` is a `ListDType`, otherwise `None`
+    pub fn as_list(&self) -> Option<&DType> {
```
Comment (gatesn) [inline vortex-dtype/src/dtype.rs]:
```suggestion
    pub fn as_list_element(&self) -> Option<&DType> {
```

**Review by gatesn — CHANGES_REQUESTED** [review]

File: vortex-array/src/array/chunked/canonical.rs (hunk 5 above)
Comment (joseph-isaacs) [reply vortex-array/src/array/chunked/canonical.rs]: I think using a list view in the future would help avoid the offset rebase

File: vortex-array/src/array/list/mod.rs (hunk 13 above)
Comment (joseph-isaacs) [reply vortex-array/src/array/list/mod.rs]: We need to check this is sorted?

File: vortex-array/src/array/list/mod.rs (hunk 9 above)
Comment (joseph-isaacs) [reply vortex-array/src/array/list/mod.rs]: We want to store the nullability, so I would say the dtype still preferred?

Comment (gatesn) [reply vortex-array/src/array/list/mod.rs]: I thought you assert that offsets are non-nullable?

File: vortex-array/src/arrow/array.rs (hunk 14 above)
Comment (joseph-isaacs) [reply vortex-array/src/arrow/array.rs:185]: ?

File: vortex-array/src/array/list/mod.rs (hunk 11 above)
Comment (joseph-isaacs) [reply vortex-array/src/array/list/mod.rs]: I added an unchecked path

File: vortex-array/src/arrow/array.rs (hunk 14 above)
Comment (gatesn) [reply vortex-array/src/arrow/array.rs:185]: `ArrayData::from_arrow(value.offsets().clone(), false)`

**Review by gatesn — APPROVED** [review]

**Pull request merged**